### PR TITLE
Buy album

### PR DESCRIPTION
**Pull request by shernanm**
## Summary

* The endpoint that allows you to buy an album is implemented.
* Their respective tests are carried out.
* Mocks are made for the response of the external service.

## Trello Card

https://trello.com/c/aUMjjbMh

### CODE REVIEW

File: app/controllers/album.js [hunk 1]
```diff
@@ -1,11 +1,13 @@
-const albumsService = require('../services/albums');
+const { getAlbums, getPhotosBy, albumRegister, findAlbumBy } = require('../services/albums');
 const logger = require('../logger');
 const message = require('../constants');
+const { decodedToken } = require('../utils');
+const { albumMapper } = require('../mappers/mappers');
+const errors = require('../errors');
 
-exports.getAlbums = (req, res, next) => {
+exports.getAllAlbums = (req, res, next) => {
```
Comment (raparicio6) [inline app/controllers/album.js]: its ok to call getAlbums in order to make the endpoint more generic. you may want to get all the albums sometimes but other times you may only want some albums passing a filter.
it wont be good to create two different endpoints for doing that

File: app/controllers/album.js [hunk 2]
```diff
@@ -18,12 +20,37 @@ exports.getPhotos = (req, res, next) => {
...
     .then(response => {
       const albumImgUrl = response.map(({ url }) => url);
       logger.info(message.MESSAGE_OK);
       return res.status(200).send(albumImgUrl);
     })
     .catch(next);
 };
+
+exports.buyAlbums = (req, res, next) => {
+  const qs = {
```
Comment (raparicio6) [inline app/controllers/album.js]: why do you do this?
you can pass req.params to findAlbumBy

File: app/controllers/album.js [hunk 3]
```diff
@@ -18,12 +20,37 @@ exports.getPhotos = (req, res, next) => {
...
+    id: req.params.id
+  };
+  logger.info(`buyAlbums method start, request methods: ${req.method}, endpoint: ${req.path}, id: ${qs.id}`);
+  findAlbumBy(qs)
```
Comment (raparicio6) [inline app/controllers/album.js]: add return

File: app/controllers/album.js [hunk 4]
```diff
@@ -18,12 +20,37 @@ exports.getPhotos = (req, res, next) => {
...
+    .then(purchasedAlbum => {
+      if (purchasedAlbum) {
+        return next(errors.buyAlbumError('Duplicate purchase of an album is not allowed'));
+      }
+      const user = decodedToken(req.headers.token);
+      return getAlbums(qs).then(([{ id, title }]) => {
```
Comment (raparicio6) [inline app/controllers/album.js]: remove { }

File: app/controllers/album.js [hunk 5]
```diff
@@ -18,12 +20,37 @@ exports.getPhotos = (req, res, next) => {
...
+        albumRegister(albumMapper(id, title, user.id)).then(() => {
+          logger.info(`Album ${title} successfully purchased`);
+          res.status(201).send({
```
Comment (raparicio6) [inline app/controllers/album.js]: add return

File: app/services/users.js [hunk 6]
```diff
@@ -69,3 +69,43 @@ exports.changeRole = (role, email) =>
...
+    offset,
+    raw: true,
+    attributes: ['id', 'firstName', 'lastName', 'email']
+  }).catch(err => {
+    logger.info('Error trying to find the user');
+    throw errors.databaseError(`${err}`);
+  });
+
+exports.AdminUserRegister = user =>
+  User.create({ ...user, role: 'admin' }).catch(() => {
```
Comment (raparicio6) [inline app/services/users.js]: why another method for admin user?
if you have 10 roles you would create 10 methods?
try to reuse the original create user

**Review by raparicio6 — CHANGES_REQUESTED** [review]

File: app/services/users.js (hunk 6 above)
Comment (shernanm) [reply app/services/users.js]: It was my mistake, I was using it but in the end reuse userRegister.

File: app/controllers/album.js (hunk 2 above)
Comment (shernanm) [reply app/controllers/album.js]: I did it so the method is general, plus I'm also using it below in getAlbums.

Comment (raparicio6) [reply app/controllers/album.js]: Its ok but not neccesary. You are just copying req.params object into a new one called qs

File: app/mappers/mappers.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+exports.mapperUserList = ({ count, rows }, limit, page) => {
```
Comment (raparicio6) [inline app/mappers/mappers.js]: dont create a big `mapper` file. 
create one file per controller

File: package.json [hunk 8]
```diff
@@ -64,9 +64,12 @@
     "jsonwebtoken-promisified": "^1.0.3",
     "jwt-simple": "^0.5.1",
     "morgan": "^1.9.1",
+    "nock": "^10.0.6",
```
Comment (raparicio6) [inline package.json]: you need this?

File: test/album.test.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
+const request = require('supertest');
+const dictum = require('dictum.js');
+const { albumMock, errorAlbumMock } = require('./mocks/albums');
+
```
Comment (raparicio6) [inline test/album.test.js:3]: why the blank lines?

File: test/mocks/albums.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+const albumResponse = [
+  {
+    id: 1,
+    title: 'quidem molestiae enim'
+  }
+];
+
+exports.albumMock = (value = albumResponse) => requestPromise.mockResolvedValue(value);
+exports.errorAlbumMock = () =>
+  requestPromise.mockRejectedValue(new Error(errors.albumError(MESSAGE_ALBUM_API_FAILED)));
```
Comment (raparicio6) [inline test/mocks/albums.js]: do you need the Error object? We never use it

**Review by raparicio6 — CHANGES_REQUESTED** [review]

**Review by molinajulian — APPROVED** [review]

**Pull request merged**